### PR TITLE
Fixed: make WriteResponseHeader goroutine safe

### PR DESCRIPTION
**Pull request by Dogild**
Previously the function WriteResponseHeader from the package api.encoding was not goroutine safe. Usage of the method for a same context with 2 goroutines would result in a panic due to concurrent write.

In this PR, a mutext is being added in `InboundCall` and used by the method `WriteResponseHeader` to protect the update of resHeaders. 

- [x] Description and context for reviewers: one partner, one stranger
- [ ] Docs (package doc)
- [ ] Entry in CHANGELOG.md

### CODE REVIEW

File: api/encoding/inbound_call.go [hunk 1]
```diff
@@ -157,6 +159,9 @@ func (ic *inboundCallMetadata) WriteResponseHeader(k, v string) error {
 	if ic.disableResponseHeaders {
 		return yarpcerrors.InvalidArgumentErrorf("call does not support setting response headers")
 	}
+
+	ic.headersMutext.Lock()
+	defer ic.headersMutext.Unlock()
 	ic.resHeaders = append(ic.resHeaders, keyValuePair{k: k, v: v})
```
Comment (jrusev) [inline api/encoding/inbound_call.go:165]: Perhaps document that the method is safe for concurrent use.

**Comment by AllenLuUber** [comment]
I think the question here is about who maintains this lock.
If concurrent write is common and it is common behavior that multiple go routines would modify this header, yarpc-go will maintain it in the core lib. Otherwise, it should be caller's responsibility to maintain the lock.

I think this concurrent write is not that common and should be caller's responsibility. Adding this on the hotpath seems expensive. Let's see original author @abhinav 's thoughts are.

File: api/encoding/inbound_call.go [hunk 2]
```diff
@@ -34,6 +35,7 @@ import (
 // incoming request on the Context and send response headers through
 // WriteResponseHeader.
 type InboundCall struct {
+	headersMutext          sync.Mutex
```
Comment (AllenLuUber) [inline api/encoding/inbound_call.go:38]: if we decide to keep this PR, use embed over a field for mutex.
```
type InboundCall struct {
          sync.Mutex
}

**Comment by Dogild** [comment]
> I think the question here is about who maintains this lock.
> If concurrent write is common and it is common behavior that multiple go routines would modify this header, yarpc-go will maintain it in the core lib. Otherwise, it should be caller's responsibility to maintain the lock.
> 
> I think this concurrent write is not that common and should be caller's responsibility. Adding this on the hotpath seems expensive. Let's see original author @abhinav 's thoughts are.



> I think the question here is about who maintains this lock.
> If concurrent write is common and it is common behavior that multiple go routines would modify this header, yarpc-go will maintain it in the core lib. Otherwise, it should be caller's responsibility to maintain the lock.
> 
> I think this concurrent write is not that common and should be caller's responsibility. Adding this on the hotpath seems expensive. Let's see original author @abhinav 's thoughts are.

Yep agree with this! Currently talking with engineers asking for this to see if they can do this logic on their side

**Comment by Dogild** [comment]
Closing this PR.

After discussion with other owners of YARPC and engineers asking for this request, we have decided to not put the lock in YARPC but let the responsibility of the users to use a lock when needed. As benefits, performances will only be affected when concurrent write is needed and not for all calls.

**Pull request closed without merging**